File: src/python/xerxes/context/headroom.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Native Headroom-style compression for model-visible tool results.

Xerxes keeps the full original in project memory, then sends the model a
bounded preview. This module provides deterministic content routing and
specialized previews for the high-noise outputs agents produce most often:
JSON, unified diffs, grep/ripgrep results, logs, and generic text.
"""

from __future__ import annotations

import json
import re
from collections import OrderedDict
from dataclasses import dataclass
from typing import Any, TypeVar

DEFAULT_HEADROOM_PREVIEW_CHARS = 4_000
_IMPORTANT_RE = re.compile(
    r"\b(error|failed?|failure|fatal|exception|traceback|warning|warn|panic|timeout|denied|segmentation|assert)\b",
    re.IGNORECASE,
)
_SUMMARY_RE = re.compile(
    r"(=+ .* =+|short test summary|tests? passed|tests? failed|passed in|failed in|npm err|cargo test|all checks passed)",
    re.IGNORECASE,
)
_T = TypeVar("_T")


@dataclass(frozen=True)
class HeadroomResult:
    """Compressed tool-output preview and metadata."""

    content_type: str
    original_chars: int
    compressed_chars: int
    original_lines: int
    compressed_lines: int
    compressed: str
    notes: tuple[str, ...] = ()

    @property
    def kept_ratio(self) -> float:
        """Return compressed/original character ratio."""
        if self.original_chars <= 0:
            return 1.0
        return min(1.0, self.compressed_chars / self.original_chars)

    def metadata_line(self) -> str:
        """Return a compact human-readable compression summary."""
        percent = self.kept_ratio * 100
        return (
            f"{self.content_type} preview: {self.original_chars:,} -> "
            f"{self.compressed_chars:,} chars ({percent:.1f}% kept)"
        )


def compress_tool_result(
    tool_name: str,
    content: Any,
    *,
    max_chars: int = DEFAULT_HEADROOM_PREVIEW_CHARS,
) -> HeadroomResult:
    """Compress a tool result into a bounded preview.

    Args:
        tool_name: Tool that produced the output. Used only as a hint.
        content: Raw tool output.
        max_chars: Maximum preview size before a final fit pass.

    Returns:
        A routed, bounded preview with metadata.
    """
    text = _stringify(content)
    if not text:
        return _result("empty", text, "", notes=("empty tool result",))
    if _looks_binary(text):
        preview = "[Binary-like tool result omitted. Read the stored project-memory artifact for bytes.]"
        return _result("binary", text, preview, notes=("binary-like content",))

    bounded_chars = max(512, int(max_chars))
    stripped = text.lstrip()
    if stripped.startswith(("{", "[")):
        compressed = _compress_json(text, bounded_chars)
        if compressed is not None:
            return _result("json", text, compressed)
    if _looks_like_diff(text):
        return _result("diff", text, _compress_diff(text, bounded_chars))
    if _looks_like_search_output(text):
        return _result("search", text, _compress_search(text, bounded_chars))
    if _looks_like_log(text, tool_name=tool_name):
        return _result("log", text, _compress_log(text, bounded_chars))
    return _result("text", text, _compress_text(text, bounded_chars))


def _result(content_type: str, original: str, compressed: str, *, notes: tuple[str, ...] = ()) -> HeadroomResult:
    fitted = compressed.rstrip()
    return HeadroomResult(
        content_type=content_type,
        original_chars=len(original),
        compressed_chars=len(fitted),
        original_lines=len(original.splitlines()),
        compressed_lines=len(fitted.splitlines()),
        compressed=fitted,
        notes=notes,
    )


def _stringify(content: Any) -> str:
    if isinstance(content, str):
        return content
    try:
        return json.dumps(content, ensure_ascii=False, indent=2, sort_keys=True)
    except (TypeError, ValueError):
        return str(content)


def _looks_binary(text: str) -> bool:
    if "\x00" in text:
        return True
    sample = text[:4096]
    if not sample:
        return False
    control_count = sum(1 for ch in sample if ord(ch) < 32 and ch not in "\n\r\t")
    return control_count / len(sample) > 0.05


def _fit_text(text: str, max_chars: int) -> str:
    if len(text) <= max_chars:
        return text.rstrip()
    marker = f"\n\n[... {len(text) - max_chars:,} chars omitted by Xerxes headroom ...]\n\n"
    remaining = max(256, max_chars - len(marker))
    head = remaining // 2
    tail = remaining - head
    return f"{text[:head].rstrip()}{marker}{text[-tail:].lstrip()}".rstrip()


def _compress_json(text: str, max_chars: int) -> str | None:
    try:
        value = json.loads(text)
    except json.JSONDecodeError:
        return None

    minified = json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    if len(minified) <= max_chars:
        return minified

    shrunk = _shrink_json(value)
    rendered = json.dumps(shrunk, ensure_ascii=False, indent=2, sort_keys=True)
    return _fit_text(rendered, max_chars)


def _shrink_json(value: Any, *, depth: int = 0) -> Any:
    if depth >= 4:
        return _json_atom_preview(value)
    if isinstance(value, dict):
        items = list(value.items())
        if len(items) > 24:
            important = [(key, item) for key, item in items if _json_key_is_important(key)]
            head = items[:10]
            tail = items[-5:]
            selected = _dedupe_pairs([*important[:8], *head, *tail])
            omitted = len(items) - len(selected)
        else:
            selected = items
            omitted = 0
        out: dict[str, Any] = {str(key): _shrink_json(item, depth=depth + 1) for key, item in selected}
        if omitted > 0:
            out["..."] = f"{omitted} keys omitted"
        return out
    if isinstance(value, list):
        important = [item for item in value if _json_value_is_important(item)]
        sample = _dedupe_json_items([*important[:5], *value[:3], *value[-2:]])
        omitted = len(value) - len(sample)
        return {
            "type": "array",
            "items": len(value),
            "sample": [_shrink_json(item, depth=depth + 1) for item in sample],
            "omitted": max(0, omitted),
        }
    return _json_atom_preview(value)


def _json_key_is_important(key: Any) -> bool:
    return bool(_IMPORTANT_RE.search(str(key)))


def _json_value_is_important(value: Any) -> bool:
    try:
        rendered = json.dumps(value, ensure_ascii=False, sort_keys=True)
    except (TypeError, ValueError):
        rendered = str(value)
    return bool(_IMPORTANT_RE.search(rendered))


def _json_atom_preview(value: Any) -> Any:
    if isinstance(value, str) and len(value) > 240:
        return f"{value[:160]} ... [{len(value) - 220} chars omitted] ... {value[-60:]}"
    return value


def _dedupe_pairs(items: list[tuple[Any, Any]]) -> list[tuple[Any, Any]]:
    seen: set[str] = set()
    selected: list[tuple[Any, Any]] = []
    for key, value in items:
        marker = str(key)
        if marker not in seen:
            seen.add(marker)
            selected.append((key, value))
    return selected


def _dedupe_json_items(items: list[Any]) -> list[Any]:
    seen: set[str] = set()
    selected: list[Any] = []
    for item in items:
        try:
            marker = json.dumps(item, ensure_ascii=False, sort_keys=True)
        except (TypeError, ValueError):
            marker = repr(item)
        if marker not in seen:
            seen.add(marker)
            selected.append(item)
    return selected


def _looks_like_diff(text: str) -> bool:
    return text.startswith("diff --git ") or "\ndiff --git " in text or bool(re.search(r"^@@ -\d+", text, re.MULTILINE))


def _compress_diff(text: str, max_chars: int) -> str:
    files = _split_diff_files(text)
    if not files:
        return _compress_text(text, max_chars)

    max_files = 12
    total_files = len(files)
    selected_files = files[:max_files]
    out: list[str] = [
        f"[Xerxes headroom diff preview: {total_files} files, showing {len(selected_files)}]",
    ]
    for file_index, file_text in enumerate(selected_files, start=1):
        reduced = _compress_diff_file(file_text)
        out.extend(["", f"# File {file_index}", reduced.rstrip()])
    if total_files > len(selected_files):
        out.append(f"\n[... {total_files - len(selected_files)} diff files omitted ...]")
    return _fit_text("\n".join(out), max_chars)


def _split_diff_files(text: str) -> list[str]:
    starts = [match.start() for match in re.finditer(r"^diff --git ", text, re.MULTILINE)]
    if not starts:
        return [text] if "@@ -" in text else []
    starts.append(len(text))
    return [text[starts[index] : starts[index + 1]].rstrip() for index in range(len(starts) - 1)]


def _compress_diff_file(file_text: str) -> str:
    lines = file_text.splitlines()
    hunk_indices = [index for index, line in enumerate(lines) if line.startswith("@@ ")]
    if not hunk_indices:
        return "\n".join(lines[:40])

    first_hunk = hunk_indices[0]
    header = lines[:first_hunk]
    hunk_bounds = [*hunk_indices, len(lines)]
    hunks = [lines[hunk_bounds[index] : hunk_bounds[index + 1]] for index in range(len(hunk_bounds) - 1)]
    selected = hunks[:4]
    if len(hunks) > 6:
        selected.extend(hunks[-2:])
    elif len(hunks) > 4:
        selected.extend(hunks[4:])

    out = [*header[:12]]
    if len(header) > 12:
        out.append(f"[... {len(header) - 12} header lines omitted ...]")
    for hunk in selected:
        out.extend(_trim_diff_hunk(hunk))
    if len(hunks) > len(selected):
        out.append(f"[... {len(hunks) - len(selected)} hunks omitted ...]")
    return "\n".join(out)


def _trim_diff_hunk(hunk: list[str], *, context: int = 2) -> list[str]:
    if not hunk:
        return []
    keep = {0}
    for index, line in enumerate(hunk):
        if index == 0:
            continue
        if (line.startswith("+") and not line.startswith("+++")) or (
            line.startswith("-") and not line.startswith("---")
        ):
            for kept_index in range(max(1, index - context), min(len(hunk), index + context + 1)):
                keep.add(kept_index)
    out: list[str] = []
    previous = -1
    for index in sorted(keep):
        if previous != -1 and index > previous + 1:
            out.append(f"[... {index - previous - 1} context lines omitted ...]")
        out.append(hunk[index])
        previous = index
    return out


def _looks_like_search_output(text: str) -> bool:
    lines = [line for line in text.splitlines()[:80] if line.strip()]
    if len(lines) < 6:
        return False
    parsed = sum(1 for line in lines if _parse_search_line(line) is not None)
    return parsed >= max(4, len(lines) // 3)


def _parse_search_line(line: str) -> tuple[str, int, str] | None:
    for match in re.finditer(r"[:\-](\d+)[:\-]", line):
        path = line[: match.start()]
        if not path:
            continue
        try:
            number = int(match.group(1))
        except ValueError:
            continue
        content = line[match.end() :]
        return path, number, content
    return None


def _compress_search(text: str, max_chars: int) -> str:
    grouped: OrderedDict[str, list[tuple[int, str]]] = OrderedDict()
    unparsed = 0
    total_matches = 0
    for line in text.splitlines():
        parsed = _parse_search_line(line)
        if parsed is None:
            unparsed += 1
            continue
        path, number, content = parsed
        grouped.setdefault(path, []).append((number, content))
        total_matches += 1

    max_files = 20
    max_per_file = 6
    out = [f"[Xerxes headroom search preview: {total_matches} matches across {len(grouped)} files]"]
    for path, matches in list(grouped.items())[:max_files]:
        shown = _select_first_last(matches, max_per_file)
        out.extend(["", path])
        for number, content in shown:
            out.append(f"  {number}: {content}")
        omitted = len(matches) - len(shown)
        if omitted > 0:
            out.append(f"  [... {omitted} more matches in this file ...]")
    dropped_files = max(0, len(grouped) - max_files)
    if dropped_files:
        out.append(f"\n[... {dropped_files} files omitted ...]")
    if unparsed:
        out.append(f"\n[... {unparsed} non-search lines omitted ...]")
    return _fit_text("\n".join(out), max_chars)


def _select_first_last(items: list[_T], limit: int) -> list[_T]:
    if len(items) <= limit:
        return list(items)
    head_count = max(1, limit // 2)
    tail_count = max(1, limit - head_count)
    return [*items[:head_count], *items[-tail_count:]]


def _looks_like_log(text: str, *, tool_name: str) -> bool:
    lowered_name = tool_name.lower()
    if any(hint in lowered_name for hint in ("exec", "shell", "test", "pytest", "build", "npm", "cargo")):
        return True
    lines = text.splitlines()
    if len(lines) < 40:
        return False
    marked = sum(1 for line in lines[:200] if _IMPORTANT_RE.search(line) or _SUMMARY_RE.search(line))
    return marked > 0


def _compress_log(text: str, max_chars: int) -> str:
    lines = text.splitlines()
    important = {
        index
        for index, line in enumerate(lines)
        if _IMPORTANT_RE.search(line) or _SUMMARY_RE.search(line) or line.startswith(("FAILED ", "ERROR "))
    }
    keep = set(range(min(8, len(lines))))
    keep.update(range(max(0, len(lines) - 10), len(lines)))
    for index in important:
        for kept_index in range(max(0, index - 2), min(len(lines), index + 3)):
            keep.add(kept_index)

    out = [f"[Xerxes headroom log preview: kept {len(keep)} of {len(lines)} lines]"]
    previous = -1
    for index in sorted(keep):
        if previous != -1 and index > previous + 1:
            out.append(f"[... {index - previous - 1} log lines omitted ...]")
        out.append(f"L{index + 1}: {lines[index]}")
        previous = index
    return _fit_text("\n".join(out), max_chars)


def _compress_text(text: str, max_chars: int) -> str:
    lines = text.splitlines()
    if len(lines) <= 60:
        return _fit_text(text, max_chars)
    selected = _select_first_last(lines, 40)
    omitted = len(lines) - len(selected)
    head_count = min(20, len(selected))
    out = [
        f"[Xerxes headroom text preview: {len(lines)} lines, {omitted} omitted]",
        *selected[:head_count],
    ]
    tail = selected[head_count:]
    if tail:
        out.extend([f"[... {omitted} middle lines omitted ...]", *tail])
    return _fit_text("\n".join(out), max_chars)

File: xerxes/src/context/headroom.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

export const DEFAULT_HEADROOM_PREVIEW_CHARS = 4_000

const IMPORTANT_RE = /\b(error|failed?|failure|fatal|exception|traceback|warning|warn|panic|timeout|denied|segmentation|assert)\b/i
const SUMMARY_RE = /(=+ .* =+|short test summary|tests? passed|tests? failed|passed in|failed in|npm err|cargo test|all checks passed)/i

export interface HeadroomResultOptions {
  readonly compressed: string
  readonly contentType: string
  readonly notes?: readonly string[]
  readonly original: string
}

/** Bounded model-visible preview while the caller retains the full tool result elsewhere. */
export class HeadroomResult {
  readonly compressed: string
  readonly compressedChars: number
  readonly compressedLines: number
  readonly contentType: string
  readonly notes: readonly string[]
  readonly originalChars: number
  readonly originalLines: number

  constructor(options: HeadroomResultOptions) {
    const preview = options.compressed.trimEnd()
    this.contentType = options.contentType
    this.originalChars = options.original.length
    this.compressedChars = preview.length
    this.originalLines = lineCount(options.original)
    this.compressedLines = lineCount(preview)
    this.compressed = preview
    this.notes = Object.freeze([...(options.notes ?? [])])
    Object.freeze(this)
  }

  get keptRatio(): number {
    return this.originalChars <= 0 ? 1 : Math.min(1, this.compressedChars / this.originalChars)
  }

  metadataLine(): string {
    return this.contentType + ' preview: ' + formatInteger(this.originalChars) + ' -> '
      + formatInteger(this.compressedChars) + ' chars (' + (this.keptRatio * 100).toFixed(1) + '% kept)'
  }
}

export interface CompressToolResultOptions {
  readonly maxChars?: number
}

/**
 * Route noisy tool output through deterministic JSON, diff, search, log, or
 * text previews. No provider call is made; callers can persist the original
 * result independently before sending this bounded preview to a model.
 */
export function compressToolResult(
  toolName: string,
  content: unknown,
  options: CompressToolResultOptions = {},
): HeadroomResult {
  const text = stringify(content)
  if (!text) return result('empty', text, '', ['empty tool result'])
  if (looksBinary(text)) {
    return result(
      'binary',
      text,
      '[Binary-like tool result omitted. Read the stored project-memory artifact for bytes.]',
      ['binary-like content'],
    )
  }
  const maxChars = normalizedMaxChars(options.maxChars)
  const stripped = text.trimStart()
  if (stripped.startsWith('{') || stripped.startsWith('[')) {
    const compressed = compressJson(text, maxChars)
    if (compressed !== undefined) return result('json', text, compressed)
  }
  if (looksLikeDiff(text)) return result('diff', text, compressDiff(text, maxChars))
  if (looksLikeSearch(text)) return result('search', text, compressSearch(text, maxChars))
  if (looksLikeLog(text, toolName)) return result('log', text, compressLog(text, maxChars))
  return result('text', text, compressText(text, maxChars))
}

function result(contentType: string, original: string, compressed: string, notes: readonly string[] = []): HeadroomResult {
  return new HeadroomResult({ contentType, original, compressed, notes })
}

function stringify(value: unknown): string {
  if (typeof value === 'string') return value
  try {
    const rendered = JSON.stringify(sortValue(value), null, 2)
    return rendered === undefined ? String(value) : rendered
  } catch {
    return String(value)
  }
}

function looksBinary(text: string): boolean {
  if (text.includes('\u0000')) return true
  const sample = text.slice(0, 4_096)
  if (!sample) return false
  let controls = 0
  for (const character of sample) {
    const code = character.codePointAt(0) ?? 0
    if (code < 32 && character !== '\n' && character !== '\r' && character !== '\t') controls += 1
  }
  return controls / sample.length > 0.05
}

function fitText(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text.trimEnd()
  const marker = '\n\n[... ' + formatInteger(text.length - maxChars) + ' chars omitted by Xerxes headroom ...]\n\n'
  const remaining = Math.max(2, maxChars - marker.length)
  const head = Math.floor(remaining / 2)
  const tail = remaining - head
  const output = text.slice(0, head).trimEnd() + marker + text.slice(-tail).trimStart()
  return output.length <= maxChars ? output.trimEnd() : output.slice(0, maxChars).trimEnd()
}

function compressJson(text: string, maxChars: number): string | undefined {
  let value: unknown
  try {
    value = JSON.parse(text)
  } catch {
    return undefined
  }
  const minified = stableJson(value)
  if (minified.length <= maxChars) return minified
  const rendered = JSON.stringify(shrinkJson(value), null, 2)
  return fitText(rendered, maxChars)
}

function shrinkJson(value: unknown, depth = 0): unknown {
  if (depth >= 4) return jsonAtomPreview(value)
  if (Array.isArray(value)) {
    const important = value.filter(jsonValueImportant)
    const sample = dedupeValues([...important.slice(0, 5), ...value.slice(0, 3), ...value.slice(-2)])
    return {
      type: 'array',
      items: value.length,
      sample: sample.map(item => shrinkJson(item, depth + 1)),
      omitted: Math.max(0, value.length - sample.length),
    }
  }
  if (value && typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>)
    let selected = entries
    let omitted = 0
    if (entries.length > 24) {
      const important = entries.filter(([key]) => jsonKeyImportant(key))
      selected = dedupePairs([...important.slice(0, 8), ...entries.slice(0, 10), ...entries.slice(-5)])
      omitted = entries.length - selected.length
    }
    const output: Record<string, unknown> = {}
    for (const [key, item] of selected) output[key] = shrinkJson(item, depth + 1)
    if (omitted > 0) output['...'] = omitted + ' keys omitted'
    return output
  }
  return jsonAtomPreview(value)
}

function jsonKeyImportant(key: string): boolean {
  return IMPORTANT_RE.test(key)
}

function jsonValueImportant(value: unknown): boolean {
  return IMPORTANT_RE.test(stableJson(value))
}

function jsonAtomPreview(value: unknown): unknown {
  if (typeof value === 'string' && value.length > 240) {
    return value.slice(0, 160) + ' ... [' + (value.length - 220) + ' chars omitted] ... ' + value.slice(-60)
  }
  return value
}

function dedupePairs(items: readonly (readonly [string, unknown])[]): Array<[string, unknown]> {
  const seen = new Set<string>()
  const output: Array<[string, unknown]> = []
  for (const [key, value] of items) {
    if (seen.has(key)) continue
    seen.add(key)
    output.push([key, value])
  }
  return output
}

function dedupeValues(items: readonly unknown[]): unknown[] {
  const seen = new Set<string>()
  const output: unknown[] = []
  for (const item of items) {
    const marker = stableJson(item)
    if (seen.has(marker)) continue
    seen.add(marker)
    output.push(item)
  }
  return output
}

function looksLikeDiff(text: string): boolean {
  return text.startsWith('diff --git ') || text.includes('\ndiff --git ') || /^@@ -\d+/m.test(text)
}

function compressDiff(text: string, maxChars: number): string {
  const files = splitDiffFiles(text)
  if (!files.length) return compressText(text, maxChars)
  const selected = files.slice(0, 12)
  const output = ['[Xerxes headroom diff preview: ' + files.length + ' files, showing ' + selected.length + ']']
  selected.forEach((file, index) => {
    output.push('', '# File ' + (index + 1), compressDiffFile(file).trimEnd())
  })
  if (files.length > selected.length) output.push('\n[... ' + (files.length - selected.length) + ' diff files omitted ...]')
  return fitText(output.join('\n'), maxChars)
}

function splitDiffFiles(text: string): string[] {
  const starts = [...text.matchAll(/^diff --git /gm)].map(match => match.index ?? 0)
  if (!starts.length) return text.includes('@@ -') ? [text] : []
  starts.push(text.length)
  const files: string[] = []
  for (let index = 0; index < starts.length - 1; index += 1) {
    const start = starts[index]
    const end = starts[index + 1]
    if (start !== undefined && end !== undefined) files.push(text.slice(start, end).trimEnd())
  }
  return files
}

function compressDiffFile(file: string): string {
  const lines = splitLines(file)
  const hunkIndexes = lines.flatMap((line, index) => line.startsWith('@@ ') ? [index] : [])
  if (!hunkIndexes.length) return lines.slice(0, 40).join('\n')
  const first = hunkIndexes[0] ?? 0
  const header = lines.slice(0, first)
  const bounds = [...hunkIndexes, lines.length]
  const hunks: string[][] = []
  for (let index = 0; index < bounds.length - 1; index += 1) {
    const start = bounds[index]
    const end = bounds[index + 1]
    if (start !== undefined && end !== undefined) hunks.push(lines.slice(start, end))
  }
  let selected = hunks.slice(0, 4)
  if (hunks.length > 6) selected = [...selected, ...hunks.slice(-2)]
  else if (hunks.length > 4) selected = [...selected, ...hunks.slice(4)]
  const output = [...header.slice(0, 12)]
  if (header.length > 12) output.push('[... ' + (header.length - 12) + ' header lines omitted ...]')
  for (const hunk of selected) output.push(...trimDiffHunk(hunk))
  if (hunks.length > selected.length) output.push('[... ' + (hunks.length - selected.length) + ' hunks omitted ...]')
  return output.join('\n')
}

function trimDiffHunk(hunk: readonly string[], context = 2): string[] {
  if (!hunk.length) return []
  const keep = new Set<number>([0])
  hunk.forEach((line, index) => {
    if (index === 0) return
    if ((line.startsWith('+') && !line.startsWith('+++')) || (line.startsWith('-') && !line.startsWith('---'))) {
      for (let current = Math.max(1, index - context); current < Math.min(hunk.length, index + context + 1); current += 1) {
        keep.add(current)
      }
    }
  })
  const output: string[] = []
  let previous = -1
  for (const index of [...keep].sort((left, right) => left - right)) {
    if (previous !== -1 && index > previous + 1) output.push('[... ' + (index - previous - 1) + ' context lines omitted ...]')
    output.push(hunk[index] ?? '')
    previous = index
  }
  return output
}

interface SearchLine {
  readonly content: string
  readonly line: number
  readonly path: string
}

function looksLikeSearch(text: string): boolean {
  const lines = splitLines(text).slice(0, 80).filter(line => Boolean(line.trim()))
  if (lines.length < 6) return false
  const parsed = lines.filter(line => parseSearchLine(line) !== undefined).length
  return parsed >= Math.max(4, Math.floor(lines.length / 3))
}

function parseSearchLine(line: string): SearchLine | undefined {
  const pattern = /[:\-](\d+)[:\-]/g
  let match: RegExpExecArray | null
  while ((match = pattern.exec(line)) !== null) {
    const path = line.slice(0, match.index)
    if (!path) continue
    const lineNumber = Number(match[1])
    if (!Number.isSafeInteger(lineNumber)) continue
    return { path, line: lineNumber, content: line.slice((match.index ?? 0) + match[0].length) }
  }
  return undefined
}

function compressSearch(text: string, maxChars: number): string {
  const groups = new Map<string, SearchLine[]>()
  let unparsed = 0
  let total = 0
  for (const line of splitLines(text)) {
    const parsed = parseSearchLine(line)
    if (!parsed) {
      unparsed += 1
      continue
    }
    const values = groups.get(parsed.path) ?? []
    values.push(parsed)
    groups.set(parsed.path, values)
    total += 1
  }
  const output = ['[Xerxes headroom search preview: ' + total + ' matches across ' + groups.size + ' files]']
  const entries = [...groups.entries()]
  for (const [path, matches] of entries.slice(0, 20)) {
    const selected = selectFirstLast(matches, 6)
    output.push('', path)
    for (const match of selected) output.push('  ' + match.line + ': ' + match.content)
    if (matches.length > selected.length) output.push('  [... ' + (matches.length - selected.length) + ' more matches in this file ...]')
  }
  if (entries.length > 20) output.push('\n[... ' + (entries.length - 20) + ' files omitted ...]')
  if (unparsed) output.push('\n[... ' + unparsed + ' non-search lines omitted ...]')
  return fitText(output.join('\n'), maxChars)
}

function looksLikeLog(text: string, toolName: string): boolean {
  const name = toolName.toLowerCase()
  if (['exec', 'shell', 'test', 'pytest', 'build', 'npm', 'cargo'].some(hint => name.includes(hint))) return true
  const lines = splitLines(text)
  if (lines.length < 40) return false
  return lines.slice(0, 200).some(line => IMPORTANT_RE.test(line) || SUMMARY_RE.test(line))
}

function compressLog(text: string, maxChars: number): string {
  const lines = splitLines(text)
  const keep = new Set<number>()
  for (let index = 0; index < Math.min(8, lines.length); index += 1) keep.add(index)
  for (let index = Math.max(0, lines.length - 10); index < lines.length; index += 1) keep.add(index)
  lines.forEach((line, index) => {
    if (!IMPORTANT_RE.test(line) && !SUMMARY_RE.test(line) && !line.startsWith('FAILED ') && !line.startsWith('ERROR ')) return
    for (let current = Math.max(0, index - 2); current < Math.min(lines.length, index + 3); current += 1) keep.add(current)
  })
  const output = ['[Xerxes headroom log preview: kept ' + keep.size + ' of ' + lines.length + ' lines]']
  let previous = -1
  for (const index of [...keep].sort((left, right) => left - right)) {
    if (previous !== -1 && index > previous + 1) output.push('[... ' + (index - previous - 1) + ' log lines omitted ...]')
    output.push('L' + (index + 1) + ': ' + (lines[index] ?? ''))
    previous = index
  }
  return fitText(output.join('\n'), maxChars)
}

function compressText(text: string, maxChars: number): string {
  const lines = splitLines(text)
  if (lines.length <= 60) return fitText(text, maxChars)
  const selected = selectFirstLast(lines, 40)
  const omitted = lines.length - selected.length
  const head = selected.slice(0, Math.min(20, selected.length))
  const tail = selected.slice(head.length)
  const output = ['[Xerxes headroom text preview: ' + lines.length + ' lines, ' + omitted + ' omitted]', ...head]
  if (tail.length) output.push('[... ' + omitted + ' middle lines omitted ...]', ...tail)
  return fitText(output.join('\n'), maxChars)
}

function selectFirstLast<T>(items: readonly T[], limit: number): T[] {
  if (items.length <= limit) return [...items]
  const head = Math.max(1, Math.floor(limit / 2))
  const tail = Math.max(1, limit - head)
  return [...items.slice(0, head), ...items.slice(-tail)]
}

function stableJson(value: unknown): string {
  try {
    return JSON.stringify(sortValue(value)) ?? String(value)
  } catch {
    return String(value)
  }
}

function sortValue(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortValue)
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value as Record<string, unknown>)
      .sort(([left], [right]) => left.localeCompare(right))
      .map(([key, item]) => [key, sortValue(item)]))
  }
  return value
}

function normalizedMaxChars(value: number | undefined): number {
  if (value === undefined) return DEFAULT_HEADROOM_PREVIEW_CHARS
  if (!Number.isFinite(value)) throw new RangeError('maxChars must be finite')
  return Math.max(512, Math.trunc(value))
}

function lineCount(value: string): number {
  return value ? splitLines(value).length : 0
}

function splitLines(value: string): string[] {
  return value.split(/\r?\n/)
}

function formatInteger(value: number): string {
  return value.toLocaleString('en-US')
}
